python/tank/context.py

```diff
@@ -624,37 +623,105 @@ def _fields_from_template_tree(self, template, fields, entities):
+                    #    Shot     | 004 | Shot_B   | /Seq_001/blah/Shot_B            <- this is out of date!
+                    #
+                    # (Note: the schema/templates have been changed since the entries for Shot_b were added)
+                    #
+                    # The sub-templates used to search for fields are:
+                    #    /{Sequence}
+                    #    /{Sequence}/{Shot}
+                    #    /{Sequence}/{Shot}/{Step}
+                    #
+                    # And the entities passed into the method are:
+                    #    Sequence:   Seq_001
+                    #    Shot:       Shot_B
+                    #    Step:       Lighting
+                    #
+                    # We are searching for fields for 'Shot_B' that has a broken entry in the path cache so the fields 
+                    # returned for each level of the template will be:
+                    #    /{Sequence}                 -> {"Sequence":"Seq_001"} <- Correct
+                    #    /{Sequence}/{Shot}          -> {}                     <- entry not found for Shot_B matching 
+                    #                                                             the template
+                    #    /{Sequence}/{Shot}/{Step}   -> {"Sequence":"Seq_001", <- Correct
+                    #                                    "Shot":"Shot_A",      <- Wrong!
+                    #                                    "Step":"Lighting"}    <- Correct
+                    #
+                    # The final fields returned are:
+                    #
+                    #     {"Sequence":"Seq_001",
+                    #      "Shot":"Shot_A",
+                    #      "Step":"Lighting"}
+                    #
+                    # The wrong Shot (Shot_A) is returned and not caught because the code only tested that the Step
+                    # entity matches and just assumes that the rest is correct - this isn't the case when there is
+                    # a one-to-many relationship between entities!
+                    #
+                    # Therefore, we need to validate that we didn't find any entity fields that we should have found
+                    # previously/higher up in the template definition.  If we did then the entries that were found 
+                    # may not be correct so we have to discard them!
+                    found_mismatching_field = False
+                    for field_name, field_value in entity_fields.iteritems():
+                        if field_name in all_fields:
+                            # already validated/processed this field so can skip
+                            continue
+                        elif field_name == key_name:
+                            # the found field matches the entity we were searching for so must be valid!
+                            found_fields[field_name] = field_value
+                        elif field_name in entities:
```

Comment:
i think i understand the logic here now, after some head scratching - not because the code is not right but because this is so tricky! I wonder if this should just be an `else`? Can you foresee cases where this would break? And in that case, should there be an `else` just for completeness? Maybe something like 

```python 
                    for field_name, field_value in entity_fields.iteritems():
                        if field_name in all_fields:
                            # the outer loop processes the template by starting
                            # at the root path and then processing downwards
                            # all identified fields are added to the all_fields list
                            # 
                            # field_name already being preset in this list indicates
                            # that this field was successfully validated against a 
                            # shorter version of this template. So this field is valid
                            # and we can skip further processing.
                            continue
                        elif field_name == key_name:
                            # the found field matches the entity we were searching for so must be valid!
                            found_fields[field_name] = field_value
                        else:
                            # We found an entity type that isn't in all_fields list - but it's not the one we are currently
                            # searching for. This must mean that it is a "left-over" that should have been found when 
                            # processing parent portions of this path. This means we can't trust any other fields that 
                            # were found as they
                            # may belong to a completely different entity/path! 
                            found_mismatching_field = True
```

Something like that... ?